### PR TITLE
Doc restructure (continued)

### PR DESCRIPTION
**Pull request by stephenfin**
Subject: Migrate API documentation into code

### Feature or Bugfix
- Enhancement

### Purpose
- Move documentation for APIs into the code, allowing us to use `help(sphinx.application.Sphinx)`.

### Detail
- As above.

### Relates
- This is a continuation of the effort started in  #4278.

### CODE REVIEW

File: sphinx/application.py [hunk 1]
```diff
@@ -461,16 +494,52 @@ def emit(self, event, *args):
...
+        The string value *rebuild* must be one of those values:
+
+        * ``'env'`` if a change in the setting only takes effect when a
+          document is parsed -- this means that the whole environment must be
+          rebuilt.
+        * ``'html'`` if a change in the setting needs a full rebuild of HTML
+          documents.
+        * ``''`` if a change in the setting will not need any special rebuild.
+
+        .. versionchanged:: 0.4
```
Comment (tk0miya) [inline sphinx/application.py]: I feel these `version***` notes are not needed for docstrings. How about moving them to the docs?

File: sphinx/application.py [hunk 2]
```diff
@@ -461,16 +494,52 @@ def emit(self, event, *args):
...
     def add_builder(self, builder):
         # type: (Type[Builder]) -> None
+        """Register a new builder.
+
+        *builder* must be a class that inherits from
+        :class:`~sphinx.builders.Builder`.
+        """
         self.registry.add_builder(builder)
 
     def add_config_value(self, name, default, rebuild, types=()):
```
Comment (tk0miya) [inline sphinx/application.py]: OMG, nobody describes the last argument... (It should be updated in another PR).

File: sphinx/application.py [hunk 3]
```diff
@@ -534,11 +644,71 @@ def add_node(self, node, **kwds):
...
+
+        Other keyword arguments are used for node visitor functions. See the
+        :meth:`Sphinx.add_node` for details.
+
+        .. versionadded:: 1.4
+        """
         self.enumerable_nodes[node] = (figtype, title_getter)
         self.add_node(node, **kwds)
 
+    # TODO(stephenfin): Remove docutils 0.4 style parsing
```
Comment (tk0miya) [inline sphinx/application.py]: Absolutely! Let's emit deprecation warnings to remove it in future.

File: sphinx/application.py [hunk 4]
```diff
@@ -534,11 +644,71 @@ def add_node(self, node, **kwds):
...
+          *final_argument_whitespace* and *option_spec* that correspond to the
+          options for the function way.  See `the Docutils docs
+          <http://docutils.sourceforge.net/docs/howto/rst-directives.html>`_
+          for details.
+
+        The directive class must inherit from the class
+        ``docutils.parsers.rst.Directive``.
+
+        For example, the (already existing) :rst:dir:`literalinclude` directive
+        would be added like this:
```
Comment (tk0miya) [inline sphinx/application.py:723]: This is an example for docutils-0.4 style. So it should be updated to newer one (in another PR).

File: sphinx/application.py [hunk 5]
```diff
@@ -561,8 +738,16 @@ def add_role(self, name, role):
...
         # type: (unicode, Any) -> None
-        # don't use roles.register_generic_role because it uses
-        # register_canonical_role
+        """Register a generic Docutils role.
+
+        Register a Docutils role that does nothing but wrap its contents in the
+        node given by *nodeclass*.
+
+        Don't use ``roles.register_generic_role`` because it uses
+        ``register_canonical_role``.
```
Comment (tk0miya) [inline sphinx/application.py]: I this this sentence is a comment for Sphinx-core developers. It will confuse the readers of document (mainly extension developers).
So please keep this as comment.

File: sphinx/application.py [hunk 6]
```diff
@@ -574,30 +759,121 @@ def add_generic_role(self, name, nodeclass):
...
+        text decoration.  If the text should be treated as literal (e.g. no
+        smart quote replacement), but not have typewriter styling, use
+        ``sphinx.addnodes.literal_emphasis`` or
+        ``sphinx.addnodes.literal_strong``.
+
+        For the role content, you have the same syntactical possibilities as
+        for standard Sphinx roles (see :ref:`xref-syntax`).
+
+        This method is also available under the deprecated alias
+        ``add_description_unit``.
```
Comment (tk0miya) [inline sphinx/application.py]: Please use `:meth:` to refer `add_description_unit`.

File: sphinx/application.py [hunk 7]
```diff
@@ -574,30 +759,121 @@ def add_generic_role(self, name, nodeclass):
...
+          object.  It will automatically add index entries if *indextemplate*
+          is nonempty; if given, it must contain exactly one instance of
+          ``%s``.  See the example below for how the template will be
+          interpreted.  * Create a new role (called *rolename*) to
+          cross-reference to these object descriptions.
+        - If you provide *parse_node*, it must be a function that takes a
+          string and a docutils node, and it must populate the node with
+          children parsed from the string.  It must then return the name of the
+          item to be used in cross-referencing and index entries.  See the
+          :file:`conf.py` file in the source for this documentation for an
```
Comment (tk0miya) [inline sphinx/application.py:861]: Note: this `:file:` role refers to `doc/conf.py`. I feel a bit strange to refer the external file here.
But I don't have good workaround for this. So let's go as is.

File: sphinx/application.py [hunk 8]
```diff
@@ -614,19 +890,67 @@ def add_description_unit(self, directivename, rolename, indextemplate='',
...
+
+        (Of course, the element following the ``topic`` directive needn't be a
+        section.)
+        """
         self.registry.add_crossref_type(directivename, rolename,
                                         indextemplate, ref_nodeclass, objname)
 
     def add_transform(self, transform):
         # type: (Type[Transform]) -> None
+        """Register a Docutils transform to be applied before parsing.
```
Comment (tk0miya) [inline sphinx/application.py]: No, this is not correct. All transforms are applied after parsing. They are called in "Reading Phase".

File: sphinx/application.py [hunk 9]
```diff
@@ -614,19 +890,67 @@ def add_description_unit(self, directivename, rolename, indextemplate='',
...
+
+        Add the standard docutils :class:`Transform` subclass *transform* to
+        the list of transforms that are applied after Sphinx parses a reST
+        document.
+        """
         self.registry.add_transform(transform)
 
     def add_post_transform(self, transform):
         # type: (Type[Transform]) -> None
+        """Register a Docutils transform to be applied after parsing.
```
Comment (tk0miya) [inline sphinx/application.py]: This is also. post_transforms are applied at "Write Phase".

File: sphinx/application.py [hunk 10]
```diff
@@ -614,19 +890,67 @@ def add_description_unit(self, directivename, rolename, indextemplate='',
...
+
+        Add the standard docutils :class:`Transform` subclass *transform* to
+        the list of transforms that are applied before Sphinx writes a
+        document.
+        """
         self.registry.add_post_transform(transform)
 
     def add_javascript(self, filename):
         # type: (unicode) -> None
+        """Register a JavaScript file to include in the default HTML template.
```
Comment (tk0miya) [inline sphinx/application.py]: "to add to HTML output" is better to me.
This does not modify the HTML "templates".

File: sphinx/application.py [hunk 11]
```diff
@@ -637,6 +961,19 @@ def add_javascript(self, filename):
 
     def add_stylesheet(self, filename, alternate=False, title=None):
         # type: (unicode, bool, unicode) -> None
+        """Register a stylesheet to include in the default HTML template.
```
Comment (tk0miya) [inline sphinx/application.py]: This is also.

File: doc/extdev/appapi.rst [hunk 12]
```diff
@@ -23,387 +23,73 @@ These methods are usually called in an extension's ``setup()`` function.
...
-   include.  Like for :meth:`add_javascript`, the filename must be relative to
-   the HTML static path, or a full URI with scheme.
-
-   .. versionadded:: 1.0
-
-   .. versionchanged:: 1.6
-      Optional ``alternate`` and/or ``title`` attributes can be supplied with
-      the *alternate* (of boolean type) and *title* (a string) arguments. The
-      default is no title and *alternate* = ``False`` (see `this explanation
-      <https://developer.mozilla.org/en-US/docs/Web/CSS/Alternative_style_sheets>`_).
```
Comment (tk0miya) [inline doc/extdev/appapi.rst:317]: It seems this link is removed. Is this intended?
I don't know alternate attribute is commonly known or not.

File: sphinx/application.py [hunk 13]
```diff
@@ -650,12 +987,34 @@ def add_stylesheet(self, filename, alternate=False, title=None):
...
+        Add *packagename* to the list of packages that LaTeX source code will
+        include.  If you provide *options*, it will be taken to `\\usepackage`
+        declaration.
+
+        .. code-block:: python
+
+           # => \\usepackage{mypackage}
+           app.add_latex_package('mypackage')
+           # => \\usepackage[foo,bar]{mypackage}
+           app.add_latex_package('mypackage', 'foo,bar')
```
Comment (tk0miya) [inline sphinx/application.py]: IMO, it is better to place result line (`# =>`) after this line.

File: sphinx/application.py [hunk 14]
```diff
@@ -664,34 +1023,83 @@ def add_lexer(self, alias, lexer):
...
         self.registry.add_documenter(cls.objtype, cls)
         self.add_directive('auto' + cls.objtype, AutodocDirective)
 
     def add_autodoc_attrgetter(self, typ, getter):
         # type: (Type, Callable[[Any, unicode, Any], Any]) -> None
+        """Register a new ``getattr``-like function for the autodoc extension.
+
+        Add *getter*, which must be a function with an interface compatible to
+        the :func:`getattr` builtin, as the autodoc attribute getter for
+        objects that are instances of *type*.  All cases where autodoc needs to
```
Comment (tk0miya) [inline sphinx/application.py]: Now we use `typ` instead. Could you update this please?

File: sphinx/application.py [hunk 15]
```diff
@@ -440,17 +459,31 @@ def import_object(self, objname, source=None):
...
+        The method returns a "listener ID" that can be used as an argument to
+        :meth:`disconnect`.
+        """
         listener_id = self.events.connect(event, callback)
         logger.debug('[app] connecting event %r: %r [id=%s]', event, callback, listener_id)
         return listener_id
 
     def disconnect(self, listener_id):
         # type: (int) -> None
+        """Unregister callback *listener_id*."""
```
Comment (tk0miya) [inline sphinx/application.py]: Unregister callback **by** ?

File: doc/extdev/appapi.rst [hunk 16]
```diff
@@ -676,8 +355,11 @@ Use this to adapt your extension to API changes in Sphinx.
 
 .. data:: version_info
```
Comment (tk0miya) [inline doc/extdev/appapi.rst]: Is there no way to use source code comment? (I'll take a look later)

File: doc/extdev/appapi.rst [hunk 17]
```diff
@@ -23,387 +23,73 @@ These methods are usually called in an extension's ``setup()`` function.
...
+.. automethod:: Sphinx.add_event(name)
 
-   Add *cls* as a new documenter class for the :mod:`sphinx.ext.autodoc`
-   extension.  It must be a subclass of :class:`sphinx.ext.autodoc.Documenter`.
-   This allows to auto-document new types of objects.  See the source of the
-   autodoc module for examples on how to subclass :class:`Documenter`.
+.. automethod:: Sphinx.set_translator(name, translator_class)
 
-   .. XXX add real docs for Documenter and subclassing
+.. automethod:: Sphinx.add_node(node, \*\*kwds)
```
Comment (tk0miya) [inline doc/extdev/appapi.rst]: Is this escape needed?

File: doc/extdev/appapi.rst [hunk 18]
```diff
@@ -23,387 +23,73 @@ These methods are usually called in an extension's ``setup()`` function.
...
-      <https://developer.mozilla.org/en-US/docs/Web/CSS/Alternative_style_sheets>`_).
-
-.. method:: Sphinx.add_latex_package(packagename, options=None)
-
-   Add *packagename* to the list of packages that LaTeX source code will include.
-   If you provide *options*, it will be taken to `\usepackage` declaration.
+.. currentmodule:: sphinx.application
 
-   .. code-block:: python
+.. automethod:: Sphinx.setup_extension(name)
```
Comment (tk0miya) [inline doc/extdev/appapi.rst]: autodoc can detect arguments of methods and functions automatically. So it might be better to drop arguments.

**Review by tk0miya — CHANGES_REQUESTED** [review]
Basically LGTM

File: sphinx/application.py (hunk 1 above)
Comment (stephenfin) [reply sphinx/application.py]: Ideally I'd like to move all this to a reference section and rework everything to use `sphinx.ext.autosummary` (it's good to dogfood our own features). As a result, I think this stuff is best here. It takes very little space.

File: doc/extdev/appapi.rst (hunk 17 above)
Comment (stephenfin) [reply doc/extdev/appapi.rst]: It is needed in order to enable proper syntax highlighting when editing the docs. It has no impact on the output.

File: doc/extdev/appapi.rst (hunk 12 above)
Comment (stephenfin) [reply doc/extdev/appapi.rst:317]: That was a mistake. I'll re-add this.

File: doc/extdev/appapi.rst (hunk 16 above)
Comment (stephenfin) [reply doc/extdev/appapi.rst]: Yes, I did add this [here](https://github.com/sphinx-doc/sphinx/pull/4468/files#diff-89b3f0fe820c207f1b0a3e78cd8ad87a).

File: sphinx/application.py (hunk 15 above)
Comment (stephenfin) [reply sphinx/application.py]: Done

File: sphinx/application.py (hunk 2 above)
Comment (stephenfin) [reply sphinx/application.py]: I added a TODO

File: sphinx/application.py (hunk 4 above)
Comment (stephenfin) [reply sphinx/application.py:723]: I extended the TODO to note this.

File: sphinx/application.py (hunk 5 above)
Comment (stephenfin) [reply sphinx/application.py]: Done

File: sphinx/application.py (hunk 6 above)
Comment (stephenfin) [reply sphinx/application.py]: Done

File: sphinx/application.py (hunk 8 above)
Comment (stephenfin) [reply sphinx/application.py]: Done

File: sphinx/application.py (hunk 9 above)
Comment (stephenfin) [reply sphinx/application.py]: Done

File: doc/extdev/appapi.rst (hunk 16 above)
Comment (tk0miya) [reply doc/extdev/appapi.rst]: Sorry, what I meant is the way to import the comment using autodoc.
I guess `autodata` or `autoattribute` work well for this.

File: doc/extdev/appapi.rst (hunk 17 above)
Comment (tk0miya) [reply doc/extdev/appapi.rst]: Okay

File: sphinx/application.py (hunk 11 above)
Comment (stephenfin) [reply sphinx/application.py]: Done

File: sphinx/application.py (hunk 13 above)
Comment (stephenfin) [reply sphinx/application.py]: Done

File: sphinx/application.py (hunk 14 above)
Comment (stephenfin) [reply sphinx/application.py]: Done

File: doc/extdev/appapi.rst (hunk 16 above)
Comment (stephenfin) [reply doc/extdev/appapi.rst]: Oh, whoops. I thought I'd done this but I clearly hadn't. Fixed in the next version.

**Review by tk0miya — APPROVED** [review]

**Comment by tk0miya** [comment]
LGTM 👍

**Review by shimizukawa — APPROVED** [review]
LGTM. 

IMO, Reference documents should be placed into docstring. Narrative documents should be placed on rst files.

**Comment by tk0miya** [comment]
Thanks!
